<launch>
  <arg name="uav_name" default="$(optenv UAV_NAME uav2)"/>

  <arg name="standalone" default="false"/>

  <arg unless="$(arg standalone)" name="nodelet" value="load"/>
  <arg     if="$(arg standalone)" name="nodelet" value="standalone"/>
  <arg unless="$(arg standalone)" name="nodelet_manager" value="$(arg uav_name)_uvdar_nodelet_manager"/>
  <arg     if="$(arg standalone)" name="nodelet_manager" value=""/>

  <arg name="threshold" default="100"/>

  <arg name="left" default="$(optenv BLUEFOX_UV_LEFT)"/>
  <arg name="right" default="$(optenv BLUEFOX_UV_RIGHT)"/>
  <arg name="left_camera_name" default="mv_$(arg left)"/>
  <arg name="right_camera_name" default="mv_$(arg right)"/>
  <arg name="expose_us_left" default="$(optenv BLUEFOX_UV_LEFT_EXPOSE_US)"/>
  <arg name="expose_us_right" default="$(optenv BLUEFOX_UV_RIGHT_EXPOSE_US)"/>

  <arg name="blink_process_rate" default="10"/>

  <arg name="accumulator_length"         default="23"/>
  <arg name="pitch_steps"         default="16"/>
  <arg name="yaw_steps"         default="16"/>
  <arg name="max_pixel_shift"         default="3"/>

  <arg name="filterDistLength"         default="10"/>
  <arg name="filterOrientationLength"         default="10"/>

  <!-- Node Settings -->
  <arg name="output" default="screen"/>
  <arg name="proc" default="false"/>
  <arg name="view" default="false"/>
  <arg name="calib" default="false"/>

  <arg name="debug" default="false"/>
  <arg name="visual_debug" default="false"/>
  <arg name="gui" default="true"/>
  <!-- <arg name="gui" default="false"/> -->

  <arg name="trajectoryControl" default="true"/>

  <arg name="followDistance" default="5"/>
  <arg name="trajCoeff" default="0.8"/>
  <arg name="yawCoeff" default="0.1"/>
  <arg name="tailingCoeff" default="1.0"/>

  <arg name="publish" default="true"/>
  <arg name="useOdom" default="false"/>
  <arg name="cemeraRotated" default="false"/>
  <arg name="TimeScale" default="1.0"/>

  <arg name="frameNatoName" default="$(optenv NATO_NAME)"/>

  <arg name="beacon" default="true"/>

  <arg name="frequency0" value="30"/>
  <arg name="frequency1" value="15"/>

  <!-- ideal angle: 1.221730476 - pi/2 -->

      <!-- <origin xyz="0.108 0 0.085" rpy="1.57079632679 4.71238898038 1.57079632679" /> -->
  <!--Usage: static_transform_publisher x y z yaw pitch roll frame_id child_frame_id  period(milliseconds)-->
<!-- OR -->
  <group ns="$(arg uav_name)">
    <node
      name="uvcam_left_tf"
      pkg="tf"
      type="static_transform_publisher"
      args="0.04  0.11 0.06 -0.3490658504 0.0 -1.57079632679 $(arg uav_name)/fcu $(arg uav_name)/uvcam_left 100"/>

    <node
      name="uvcam_right_tf"
      pkg="tf"
      type="static_transform_publisher"
      args="0.04 -0.11 0.06 -2.792526803 0.0 -1.57079632679 $(arg uav_name)/fcu $(arg uav_name)/uvcam_right 100"/>

    <!-- <node pkg="nodelet" type="nodelet" name="$(arg uav_name)_uvdar_nodelet_manager" args="manager" output="screen" > -->
    <node pkg="nodelet" type="nodelet" name="$(arg uav_name)_uvdar_nodelet_manager" args="manager" output="screen" launch-prefix="debug_roslaunch">
      <param name="num_worker_threads" value="8" />
    </node>

    <node name="bluefox_emulator" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARBluefoxEmulator $(arg nodelet_manager)" output="screen" respawn="false" launch-prefix="bash -c 'sleep 5; $0 $@'">
      <rosparam param="camera_output_topics"> ["camera_left", "camera_right"] </rosparam>
      <rosparam param="calib_files"> ["default", "default"] </rosparam>
      <remap from="~camera_left" to="/$(arg uav_name)/uvdar_bluefox_left/image_raw"/>
      <remap from="~camera_right" to="/$(arg uav_name)/uvdar_bluefox_right/image_raw"/>
      <remap from="~camera_left_transfer" to="/gazebo/$(arg uav_name)/uvdar_bluefox_left/image_raw"/>
      <remap from="~camera_right_transfer" to="/gazebo/$(arg uav_name)/uvdar_bluefox_right/image_raw"/>
    </node>



    <node name="uv_detect" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARDetector $(arg nodelet_manager)" output="screen" respawn="false"> 
    <!-- <node name="uv_detect" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARDetector $(arg nodelet_manager)" output="screen" respawn="true" launch-prefix="debug_roslaunch"> --> 
      <!-- <node name="uv_detect" pkg="uvdar" type="uv_detector_node" output="screen" launch-prefix="urxvt -e gdb -q -x /home/viktor/gdb.cmds -/-args"> -->
        <!-- <node name="uv_detect" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARDetector $(arg nodelet_manager)" output="screen" respawn="true" launch-prefix="valgrind -/-tool=callgrind -/-callgrind-out-file=/home/viktor/callgrind.out -/-instr-atstart=no -/-collect-atstart=yes"> --> 

      <param name="uav_name"		type="string" value="$(arg uav_name)"/> 
      <param name="debug"		type="bool" value="$(arg debug)"/> 
      <!-- <param name="gui"             type="bool" value="true"/> -->
      <param name="gui"             type="bool" value="false"/>
      <!-- <param name="publish_visualization"             type="bool" value="true"/> -->
      <param name="publish_visualization"             type="bool" value="false"/>
      <param name="publish"         type="bool" value="$(arg publish)"/>
      <param name="justReport" type="bool" value="true"/>
      <param name="threshold" type="int" value="$(arg threshold)"/>

      <rosparam param="camera_topics"> ["camera_left", "camera_right"] </rosparam>
      <rosparam param="points_seen_topics"> ["points_seen_left", "points_seen_right"] </rosparam>

      <remap from="~camera_left" to="/$(arg uav_name)/uvdar_bluefox_left/image_raw"/>
      <remap from="~camera_right" to="/$(arg uav_name)/uvdar_bluefox_right/image_raw"/>
      <remap from="~points_seen_left" to="/$(arg uav_name)/uvdar/points_seen_left"/>
      <remap from="~points_seen_right" to="/$(arg uav_name)/uvdar/points_seen_right"/>
      <remap from="~points_seen_left/sun" to="/$(arg uav_name)/uvdar/points_seen_left/sun"/>
      <remap from="~points_seen_right/sun" to="/$(arg uav_name)/uvdar/points_seen_right/sun"/>

      <remap from="~odometry" to="/$(arg uav_name)/mrs_odometry/new_odom"/>
      <remap from="~imu" to="mavros/imu/data"/>
    </node>

    <node name="blink_processor" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARBlinkProcessor $(arg nodelet_manager)" output="screen" respawn="true"> 
    <!-- <node name="blink_processor" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARBlinkProcessor $(arg nodelet_manager)" output="screen" respawn="true" launch-prefix="roslaunch_debug"> --> 
      <!-- <node name="blink_processor" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARBlinkProcessor $(arg nodelet_manager)" output="screen" respawn="true" launch-prefix="valgrind -/-tool=callgrind -/-callgrind-out-file=/home/viktor/callgrind.out -/-instr-atstart=no -/-collect-atstart=yes"> --> 
      <param name="uav_name" type = "string" value="$(arg uav_name)"/>
      <!-- <param name="debug"		type="bool" value="$(arg debug)"/> --> 
      <param name="debug"		type="bool" value="false"/> 
      <param name="visual_debug"		type="bool" value="$(arg visual_debug)"/> 
      <param name="GUI"             type="bool" value="true"/>
      <!-- <param name="GUI"             type="bool" value="$(arg gui)"/> -->
      <param name="InvertedPoints"    type="bool" value="false"/>

      <param name="beacon" type="bool" value="$(arg beacon)"/>
      <rosparam param="frequencies" subst_value="true"> [$(arg frequency0), $(arg frequency1)] </rosparam>

      <param name="blink_process_rate"         type="int" value="$(arg blink_process_rate)"/>

      <param name="accumulator_length"         type="int" value="$(arg accumulator_length)"/>
      <param name="pitch_steps"         type="int" value="$(arg pitch_steps)"/>
      <param name="yaw_steps"         type="int" value="$(arg yaw_steps)"/>
      <param name="max_pixel_shift"         type="int" value="$(arg max_pixel_shift)"/>
      <param name="nullify_radius"         type="int" value="8"/>
      <param name="use_camera_for_visualization"         type="bool" value="false"/>

      <rosparam param="camera_topics"> ["camera_left", "camera_right"] </rosparam>
      <rosparam param="points_seen_topics"> ["points_seen_left", "points_seen_right"] </rosparam>
      <rosparam param="blinkersSeenTopics"> ["blinkers_seen_left", "blinkers_seen_right"] </rosparam>
      <rosparam param="estimatedFramerateTopics"> ["estimated_framerate_left", "estimated_framerate_right"] </rosparam>

      <remap from="~camera_left" to="/$(arg uav_name)/uvdar_bluefox_left/image_raw"/>
      <remap from="~camera_right" to="/$(arg uav_name)/uvdar_bluefox_right/image_raw"/>
      <remap from="~points_seen_left" to="/$(arg uav_name)/uvdar/points_seen_left"/>
      <remap from="~points_seen_right" to="/$(arg uav_name)/uvdar/points_seen_right"/>
      <remap from="~blinkers_seen_left" to="/$(arg uav_name)/uvdar/blinkers_seen_left"/>
      <remap from="~blinkers_seen_right" to="/$(arg uav_name)/uvdar/blinkers_seen_right"/>
      <remap from="~estimated_framerate_left" to="/$(arg uav_name)/uvdar/estimated_framerate_left"/>
      <remap from="~estimated_framerate_right" to="/$(arg uav_name)/uvdar/estimated_framerate_right"/>
      <remap from="~visualization" to="/$(arg uav_name)/uvdar/blink_visualization/image_raw"/>

    </node>
      
    <!-- <node name="uvdar_uav_pose_calculator_node" pkg="uvdar" type="uvdar_uav_pose_calculator_node" output="screen" launch-prefix="debug_roslaunch"> --> 
    <node name="uvdar_uav_pose_calculator_node" pkg="uvdar" type="uvdar_uav_pose_calculator_node" output="screen" > 
      <param name="uav_name" type = "string" value="$(arg uav_name)"/>
      <param name="debug"		type="bool" value="false"/> 
      <!-- <param name="gui"             type="bool" value="$(arg gui)"/> -->
      <param name="gui"             type="bool" value="false"/>
      <param name="publish"         type="bool" value="false"/>
      <param name="frequenciesPerTarget" type="int" value="2"/>
      <param name="targetCount" type="int" value="1"/>
      <rosparam param="frequencies" subst_value="true"> [$(arg frequency0), $(arg frequency1)] </rosparam>

      <param name="quadrotor" type="bool" value="true"/>
      <param name="beacon" type="bool" value="$(arg beacon)"/>
      <param name="arm_length" type="double" value="0.245"/>
      <param name="beacon_height" type="double" value="0.2"/>

      <rosparam param="blinkersSeenTopics"> ["blinkers_seen_left", "blinkers_seen_right"] </rosparam>
      <rosparam param="estimatedFramerateTopics"> ["estimated_framerate_left", "estimated_framerate_right"] </rosparam>
      <rosparam param="cameraFrames" subst_value="true"> ["$(arg uav_name)/uvcam_left", "$(arg uav_name)/uvcam_right"] </rosparam>

      <remap from="~blinkers_seen_left" to="/$(arg uav_name)/uvdar/blinkers_seen_left"/>
      <remap from="~blinkers_seen_right" to="/$(arg uav_name)/uvdar/blinkers_seen_right"/>
      <remap from="~estimated_framerate_left" to="/$(arg uav_name)/uvdar/estimated_framerate_left"/>
      <remap from="~estimated_framerate_right" to="/$(arg uav_name)/uvdar/estimated_framerate_right"/>

      <rosparam param="calibFiles" subst_value="true"> ["calib_results_bf_uv_fe.txt", "calib_results_bf_uv_fe.txt"] </rosparam>

      <remap from="~measuredPoses1" to="/$(arg uav_name)/uvdar/measuredPosesL"/>
      <remap from="~measuredPoses2" to="/$(arg uav_name)/uvdar/measuredPosesR"/>
    </node>
  </group>


</launch>
